cmd/bashbrew/docker.go

```diff
@@ -237,16 +237,55 @@ func (r Repo) dockerBuildUniqueBits(entry *manifest.Manifest2822Entry) ([]string
 	return uniqueBits, nil
 }
 
-func dockerBuild(tag string, file string, context io.Reader, extraEnv []string) error {
+func dockerBuild(tag string, file string, context io.Reader, platform string) error {
 	args := []string{"build", "--tag", tag, "--file", file, "--rm", "--force-rm"}
 	args = append(args, "-")
 	cmd := exec.Command("docker", args...)
-	if extraEnv != nil {
-		cmd.Env = append(os.Environ(), extraEnv...)
+	if platform != "" {
+		// https://github.com/docker/cli/blob/v20.10.7/cli/command/image/build.go#L163
+		cmd.Env = append(os.Environ(), "DOCKER_DEFAULT_PLATFORM="+platform)
+		// ideally, we would set this via an explicit "--platform" flag on "docker build", but it's not supported without buildkit until 20.10+ and this is a trivial way to get Docker to do the right thing in both cases without explicitly trying to detect whether we're on 20.10+
+
 		if debugFlag {
-			fmt.Printf("$ export %q\n", extraEnv)
+			fmt.Printf("$ export DOCKER_DEFAULT_PLATFORM=%q\n", platform)
+		}
+	}
+	cmd.Stdin = context
+	if debugFlag {
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		fmt.Printf("$ docker %q\n", args)
+		return cmd.Run()
+	} else {
+		buf := &bytes.Buffer{}
+		cmd.Stdout = buf
+		cmd.Stderr = buf
+		err := cmd.Run()
+		if err != nil {
+			err = cli.NewMultiError(err, fmt.Errorf(`docker %q output:%s`, args, "\n"+buf.String()))
 		}
+		return err
 	}
+}
+
+func dockerBuildx(tag string, file string, context io.Reader, platform string, caches []string) error {
+	// TODO: use makeGitContext when docker builder supports subdirectories
+	args := []string{
+		"buildx",
+		"build",
+		"--progress", "plain",
+		"--tag", tag,
+		"--file", file,
+		"--cache-to", "type=inline",
+	}
+	for _, cache := range caches {
+		args = append(args, "--cache-from", "type=registry,ref="+cache)
```

Comment:
So I don't love how I've done this here. Essentially in the current implementation I've told buildkit to extract cache from all the tags of that build previously - which could be a few. 
I can't use the `bashbrew/cache:digest` tag to import cache from, since that changes on each update to the `GitCommit` - it there a stable place that I could easily access "the last build of this manifest entry"?

Reply:
In some other experiments, I've hacked around this by "guessing" a stable tag from the list of tags - it feels like it might be nice to have a stable `Name` field for each entry that attempts to stay consistent between updates and builds (even if the tags are changing).